### PR TITLE
Refactor modules

### PR DESCRIPTION
**Pull request by danielrbradley**
- Move lots out from schema into smaller modules with less dependencies.
- Fix a few compiler warnings along the way.
- Rename the gen `context` struct so it doesn't block us using the `context` std lib.
- Fix flakey lambda update integration test.

This was extracted from https://github.com/pulumi/pulumi-aws-native/pull/1406

### CODE REVIEW

File: examples/lambda-update/step1/index.ts [hunk 1]
```diff
@@ -31,7 +31,6 @@ export const graphqlPublicZip = new aws.s3.BucketObject(
   "lambda-source",
   {
     bucket: lambdaStore.id,
-    key: "samplefn",
```
Comment (mjeffryes) [inline examples/lambda-update/step1/index.ts:34]: Why the change to this example?

File: provider/pkg/resources/validate.go [hunk 2]
```diff
@@ -1,15 +1,17 @@
 // Copyright 2016-2021, Pulumi Corporation.
 
-package schema
+package resources
 
 import (
 	"fmt"
+	"math"
+	"strings"
```
Comment (mjeffryes) [inline provider/pkg/resources/validate.go:8]: I don't see where these we're added in the code. Were they just missing before?

File: provider/pkg/schema/gen.go [hunk 3]
```diff
@@ -1442,3 +1446,5 @@ func GatherSemantics(schemaDir string) (SemanticsSpecDocument, error) {
 	}
 	return semanticsDocument, nil
 }
+
+var ExtensionResourceToken string = "aws-native:index:ExtensionResource"
```
Comment (mjeffryes) [inline provider/pkg/schema/gen.go:1450]: I don't see where this moved from?

**Review by mjeffryes — APPROVED** [review]
Few nits, but looks good overall. Thanks for splitting this out.

File: examples/lambda-update/step1/index.ts (hunk 1 above)
Comment (danielrbradley) [reply examples/lambda-update/step1/index.ts:34]: Both BucketObjects were created with the same key in the same bucket, so one overwrote the other. Removing the key means we'll just use auto-naming based on the Pulumi identifier. 

This was surfacing as a refresh diff sometimes during the test - where the version id of the object had changed. Essentially this is multiple data races: which one gets written first or second, then once written if one gets read before or after the other has also been written - and therefore if it actually gets a different version to what it puts 🤦

Comment (mjeffryes) [reply examples/lambda-update/step1/index.ts:34]: Whoops that is bad. Thanks for clarifying!

File: provider/pkg/schema/gen.go (hunk 3 above)
Comment (danielrbradley) [reply provider/pkg/schema/gen.go:1450]: Ah, it's in the metadata.go .. and it's still there at this point, then it should be removed in the next PR. I also couldn't find where it had gone so fixed the compile error with this line.
